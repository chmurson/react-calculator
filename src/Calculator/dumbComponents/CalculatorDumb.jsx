import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { OPERATIONS } from './../operations'
import { OperationGridButton } from './OperationGridButton'
import { InputButton, InputGridButton } from './InputButtons'
import { TextResizer } from './TextResizer'

import './Calculator.styl'

const MAX_FONT_SIZE = 48
const MIN_FONT_SIZE = 12

export class CalculatorDumb extends React.Component {
  static propTypes = {
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    /**
     * @type {function(operation:{})}
     */
    onInput: PropTypes.func,
    /**
     * @type {function(input: String)}
     */
    onOperation: PropTypes.func,

    className: PropTypes.string
  }

  static defaultProps = {
    onInput: () => null,
    onOperation: () => null,
    className: ''
  }

  state = {
    fontSize: MAX_FONT_SIZE,
    maxWidth: 0
  }

  setRef = ref => {
    this.ref = ref
    this.setState({ maxWidth: ref.offsetWidth })
  }

  render() {
    return (
      <div className={classNames('calculator', this.props.className)} ref={this.setRef}>
        <div className="calculator__area-display">
          <TextResizer maxWidth={this.state.maxWidth} maxFontSize={MAX_FONT_SIZE} minFontSize={MIN_FONT_SIZE}>
            {({ measureRef }) => (
              <div ref={measureRef} className="calculator__current-input" style={{ fontSize: this.state.fontSize }}>
                {this.props.value}
              </div>
            )}
          </TextResizer>
        </div>
        <div className="calculator__area-buttons">
          <InputGridButton input="7" col={1} row={1} onClick={this.props.onInput} />
          <InputGridButton input="8" col={2} row={1} onClick={this.props.onInput} />
          <InputGridButton input="9" col={3} row={1} onClick={this.props.onInput} />
          <OperationGridButton operation={OPERATIONS.ADD} col={4} row={1} onClick={this.props.onOperation} />
          <InputGridButton input="4" col={1} row={2} onClick={this.props.onInput} />
          <InputGridButton input="5" col={2} row={2} onClick={this.props.onInput} />
          <InputGridButton input="6" col={3} row={2} onClick={this.props.onInput} />
          <OperationGridButton operation={OPERATIONS.SUB} col={4} row={2} onClick={this.props.onOperation} />
          <InputGridButton input="1" col={1} row={3} onClick={this.props.onInput} />
          <InputGridButton input="2" col={2} row={3} onClick={this.props.onInput} />
          <InputGridButton input="3" col={3} row={3} onClick={this.props.onInput} />
          <OperationGridButton operation={OPERATIONS.CALCULATE} col={4} row="3 / 5" onClick={this.props.onOperation} />
          <OperationGridButton operation={OPERATIONS.AC} col={1} row={4} onClick={this.props.onOperation} />
          <InputGridButton input="0" col={2} row={4} onClick={this.props.onInput} />
          <InputGridButton input="." col={3} row={4} onClick={this.props.onInput} />
          <InputButton style={{ display: 'none' }} input="Backspace" onClick={this.props.onInput} />
        </div>
      </div>
    )
  }
}
