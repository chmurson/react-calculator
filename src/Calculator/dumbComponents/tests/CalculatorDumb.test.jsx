import React from 'react'
import { mount } from 'enzyme'
import { CalculatorDumb } from '../CalculatorDumb'

describe('CalculatorDumb', () => {
  let component
  beforeAll(() => {
    component = mount(<CalculatorDumb value="0">text</CalculatorDumb>)
  })

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot()
  })
})
