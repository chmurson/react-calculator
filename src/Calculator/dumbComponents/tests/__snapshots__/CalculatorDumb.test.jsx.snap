// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorDumb matches snapshot 1`] = `
<CalculatorDumb
  className=""
  onInput={[Function]}
  onOperation={[Function]}
  value="0"
>
  <div
    className="calculator"
  >
    <div
      className="calculator__area-display"
    >
      <TextResizer
        maxFontSize={48}
        maxWidth={0}
        minFontSize={12}
      >
        <_class
          bounds={true}
          innerRef={[Function]}
          onResize={[Function]}
        >
          <Measure
            bounds={true}
            contentRect={
              Object {
                "bounds": Object {},
                "client": Object {},
                "entry": Object {},
                "margin": Object {},
                "offset": Object {},
                "scroll": Object {},
              }
            }
            measure={[Function]}
            measureRef={[Function]}
          >
            <div
              className="calculator__current-input"
              style={
                Object {
                  "fontSize": 48,
                }
              }
            >
              0
            </div>
          </Measure>
        </_class>
      </TextResizer>
    </div>
    <div
      className="calculator__area-buttons"
    >
      <InputGridButton
        col={1}
        input="7"
        onClick={[Function]}
        row={1}
      >
        <withGridCell(withKeyMap(Button)})
          col={1}
          keyToMap="7"
          onClick={[Function]}
          row={1}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="7"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 1,
                "gridRow": 1,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 1,
                  "gridRow": 1,
                }
              }
              title="Shortcut: 7"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 1,
                    "gridRow": 1,
                  }
                }
                tabIndex={-1}
                title="Shortcut: 7"
              >
                7
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </InputGridButton>
      <InputGridButton
        col={2}
        input="8"
        onClick={[Function]}
        row={1}
      >
        <withGridCell(withKeyMap(Button)})
          col={2}
          keyToMap="8"
          onClick={[Function]}
          row={1}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="8"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 2,
                "gridRow": 1,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 2,
                  "gridRow": 1,
                }
              }
              title="Shortcut: 8"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 2,
                    "gridRow": 1,
                  }
                }
                tabIndex={-1}
                title="Shortcut: 8"
              >
                8
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </InputGridButton>
      <InputGridButton
        col={3}
        input="9"
        onClick={[Function]}
        row={1}
      >
        <withGridCell(withKeyMap(Button)})
          col={3}
          keyToMap="9"
          onClick={[Function]}
          row={1}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="9"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 3,
                "gridRow": 1,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 3,
                  "gridRow": 1,
                }
              }
              title="Shortcut: 9"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 3,
                    "gridRow": 1,
                  }
                }
                tabIndex={-1}
                title="Shortcut: 9"
              >
                9
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </InputGridButton>
      <OperationGridButton
        col={4}
        onClick={[Function]}
        operation="ADD"
        row={1}
      >
        <withGridCell(withKeyMap(Button)})
          col={4}
          keyToMap="+"
          onClick={[Function]}
          row={1}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="+"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 4,
                "gridRow": 1,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 4,
                  "gridRow": 1,
                }
              }
              title="Shortcut: +"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 4,
                    "gridRow": 1,
                  }
                }
                tabIndex={-1}
                title="Shortcut: +"
              >
                +
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </OperationGridButton>
      <InputGridButton
        col={1}
        input="4"
        onClick={[Function]}
        row={2}
      >
        <withGridCell(withKeyMap(Button)})
          col={1}
          keyToMap="4"
          onClick={[Function]}
          row={2}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="4"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 1,
                "gridRow": 2,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 1,
                  "gridRow": 2,
                }
              }
              title="Shortcut: 4"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 1,
                    "gridRow": 2,
                  }
                }
                tabIndex={-1}
                title="Shortcut: 4"
              >
                4
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </InputGridButton>
      <InputGridButton
        col={2}
        input="5"
        onClick={[Function]}
        row={2}
      >
        <withGridCell(withKeyMap(Button)})
          col={2}
          keyToMap="5"
          onClick={[Function]}
          row={2}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="5"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 2,
                "gridRow": 2,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 2,
                  "gridRow": 2,
                }
              }
              title="Shortcut: 5"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 2,
                    "gridRow": 2,
                  }
                }
                tabIndex={-1}
                title="Shortcut: 5"
              >
                5
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </InputGridButton>
      <InputGridButton
        col={3}
        input="6"
        onClick={[Function]}
        row={2}
      >
        <withGridCell(withKeyMap(Button)})
          col={3}
          keyToMap="6"
          onClick={[Function]}
          row={2}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="6"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 3,
                "gridRow": 2,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 3,
                  "gridRow": 2,
                }
              }
              title="Shortcut: 6"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 3,
                    "gridRow": 2,
                  }
                }
                tabIndex={-1}
                title="Shortcut: 6"
              >
                6
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </InputGridButton>
      <OperationGridButton
        col={4}
        onClick={[Function]}
        operation="SUB"
        row={2}
      >
        <withGridCell(withKeyMap(Button)})
          col={4}
          keyToMap="-"
          onClick={[Function]}
          row={2}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="-"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 4,
                "gridRow": 2,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 4,
                  "gridRow": 2,
                }
              }
              title="Shortcut: -"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 4,
                    "gridRow": 2,
                  }
                }
                tabIndex={-1}
                title="Shortcut: -"
              >
                -
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </OperationGridButton>
      <InputGridButton
        col={1}
        input="1"
        onClick={[Function]}
        row={3}
      >
        <withGridCell(withKeyMap(Button)})
          col={1}
          keyToMap="1"
          onClick={[Function]}
          row={3}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="1"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 1,
                "gridRow": 3,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 1,
                  "gridRow": 3,
                }
              }
              title="Shortcut: 1"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 1,
                    "gridRow": 3,
                  }
                }
                tabIndex={-1}
                title="Shortcut: 1"
              >
                1
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </InputGridButton>
      <InputGridButton
        col={2}
        input="2"
        onClick={[Function]}
        row={3}
      >
        <withGridCell(withKeyMap(Button)})
          col={2}
          keyToMap="2"
          onClick={[Function]}
          row={3}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="2"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 2,
                "gridRow": 3,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 2,
                  "gridRow": 3,
                }
              }
              title="Shortcut: 2"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 2,
                    "gridRow": 3,
                  }
                }
                tabIndex={-1}
                title="Shortcut: 2"
              >
                2
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </InputGridButton>
      <InputGridButton
        col={3}
        input="3"
        onClick={[Function]}
        row={3}
      >
        <withGridCell(withKeyMap(Button)})
          col={3}
          keyToMap="3"
          onClick={[Function]}
          row={3}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="3"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 3,
                "gridRow": 3,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 3,
                  "gridRow": 3,
                }
              }
              title="Shortcut: 3"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 3,
                    "gridRow": 3,
                  }
                }
                tabIndex={-1}
                title="Shortcut: 3"
              >
                3
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </InputGridButton>
      <OperationGridButton
        col={4}
        onClick={[Function]}
        operation="CALCULATE"
        row="3 / 5"
      >
        <withGridCell(withKeyMap(Button)})
          col={4}
          keyToMap={
            Array [
              "=",
              "Enter",
            ]
          }
          onClick={[Function]}
          row="3 / 5"
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap={
              Array [
                "=",
                "Enter",
              ]
            }
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 4,
                "gridRow": "3 / 5",
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 4,
                  "gridRow": "3 / 5",
                }
              }
              title="Shortcut: = or Enter"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 4,
                    "gridRow": "3 / 5",
                  }
                }
                tabIndex={-1}
                title="Shortcut: = or Enter"
              >
                =
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </OperationGridButton>
      <OperationGridButton
        col={1}
        onClick={[Function]}
        operation="AC"
        row={4}
      >
        <withGridCell(withKeyMap(Button)})
          col={1}
          keyToMap="c"
          onClick={[Function]}
          row={4}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="c"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 1,
                "gridRow": 4,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 1,
                  "gridRow": 4,
                }
              }
              title="Shortcut: c"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 1,
                    "gridRow": 4,
                  }
                }
                tabIndex={-1}
                title="Shortcut: c"
              >
                AC
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </OperationGridButton>
      <InputGridButton
        col={2}
        input="0"
        onClick={[Function]}
        row={4}
      >
        <withGridCell(withKeyMap(Button)})
          col={2}
          keyToMap="0"
          onClick={[Function]}
          row={4}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="0"
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 2,
                "gridRow": 4,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 2,
                  "gridRow": 4,
                }
              }
              title="Shortcut: 0"
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 2,
                    "gridRow": 4,
                  }
                }
                tabIndex={-1}
                title="Shortcut: 0"
              >
                0
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </InputGridButton>
      <InputGridButton
        col={3}
        input="."
        onClick={[Function]}
        row={4}
      >
        <withGridCell(withKeyMap(Button)})
          col={3}
          keyToMap="."
          onClick={[Function]}
          row={4}
          style={Object {}}
        >
          <withKeyMap(Button)}
            innerRef={[Function]}
            keyToMap="."
            onClick={[Function]}
            style={
              Object {
                "gridColumn": 3,
                "gridRow": 4,
              }
            }
          >
            <Button
              buttonRef={[Function]}
              className=""
              onClick={[Function]}
              style={
                Object {
                  "gridColumn": 3,
                  "gridRow": 4,
                }
              }
              title="Shortcut: ."
            >
              <button
                className="custom-button custom-button--regular"
                onClick={[Function]}
                style={
                  Object {
                    "gridColumn": 3,
                    "gridRow": 4,
                  }
                }
                tabIndex={-1}
                title="Shortcut: ."
              >
                .
              </button>
            </Button>
          </withKeyMap(Button)}>
        </withGridCell(withKeyMap(Button)})>
      </InputGridButton>
      <InputButton
        input="Backspace"
        onClick={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
      >
        <withKeyMap(Button)}
          innerRef={[Function]}
          keyToMap="Backspace"
          onClick={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
        >
          <Button
            buttonRef={[Function]}
            className=""
            onClick={[Function]}
            style={
              Object {
                "display": "none",
              }
            }
            title="Shortcut: B or a or c or k or s or p or a or c or e"
          >
            <button
              className="custom-button custom-button--regular"
              onClick={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              tabIndex={-1}
              title="Shortcut: B or a or c or k or s or p or a or c or e"
            >
              Backspace
            </button>
          </Button>
        </withKeyMap(Button)}>
      </InputButton>
    </div>
  </div>
</CalculatorDumb>
`;
